File: Algorithms/p104_Maximum_Depth_of_Binary_Tree_1.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def maxDepth(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        self.max_depth = 0
        if root:
            self.walk(root,1)
        return self.max_depth
        
    def walk(self,p,cur_depth):
        if cur_depth > self.max_depth:
            self.max_depth = cur_depth
        if p.left:
            self.walk(p.left,cur_depth+1)
        if p.right:
            self.walk(p.right,cur_depth+1)

File: Algorithms/p104_Maximum_Depth_of_Binary_Tree/p104_Maximum_Depth_of_Binary_Tree_1.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    int maxDepth(TreeNode* root) {
        int max_depth = 0;
        if(root){
            walk(root,1,max_depth);
        }
        return max_depth;
    }
    void walk(TreeNode* p, int cur_depth, int& max_depth){
        if(cur_depth>max_depth){
            max_depth = cur_depth;
        }
        if(p->left){
            walk(p->left,cur_depth+1,max_depth);
        }
        if(p->right){
            walk(p->right,cur_depth+1,max_depth);
        }
    }
};
